File: С-PlusPlus/С-PlusPlus/Ball.cpp
#include "Main.h"

Ball::Ball(int bWidth, int bHeight)
{
	isPop = false;
	isHit = 0;
	isPaused = true;
	isStarted = true;
	velocity = { 0,0 };
	mPosX = (screenWidth - bWidth) / 2;
	mPosY = (screenHeight - bHeight) / 2;
	mCollider.w = bWidth;
	mCollider.h = bHeight;
	boost = 1;
}

void Ball::move(SDL_Rect& pCol, SDL_Rect& eCol)
{
	if (isStarted && !isPaused)
	{
		RandomLand();
		isStarted = false;
	}

	ballCentre = mPosY + (mCollider.h / 2);

	if (checkCollision(mCollider, pCol) || checkCollision(mCollider, eCol))
	{
		boost += 0.2f;
		velocity.x = -velocity.x * boost;
		velocity.y *= boost;
		isPop = true;
	}

	mPosX += velocity.x;
	mPosY += velocity.y;

	if ((mPosX <= -100) || (mPosX + mCollider.w >= screenWidth + 100))
	{
		if (mPosX <= -100)
			isHit = 2;
		if (mPosX + mCollider.w >= screenWidth + 100)
			isHit = 1;
		mPosX = (screenWidth - mCollider.w) / 2;
		mPosY = (screenHeight - mCollider.h) / 2;
		boost = 1;
		isPaused = true;
		isStarted = true;
		velocity = { 0,0 };
	}

	if ((mPosY < 0) || (mPosY + mCollider.h > screenHeight))
	{
		velocity.y = -velocity.y;
	}

	mCollider.x = mPosX;
	mCollider.y = mPosY;
}

void Ball::render()
{
	sBall.render(mPosX, mPosY);
}

void Ball::RandomLand()
{
	srand(time(NULL));
	int random = rand() % 4;
	if (random == 0)
	{
		velocity.x = 3;
		velocity.y = 3;
	}
	if (random == 1)
	{
		velocity.x = -3;
		velocity.y = 3;
	}
	if (random == 2)
	{
		velocity.x = -3;
		velocity.y = -3;
	}
	if (random == 3)
	{
		velocity.x = 3;
		velocity.y = -3;
	}
}

int Ball::getCentre()
{
	return ballCentre;
}

File: C-Sharp (MonoGame)/C-Sharp (MonoGame)/Assets/Ball.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace C_Sharp__MonoGame_
{
    class Ball : Sprite
    {
        public int Centre;
        Vector2 velocity;
        private float boost = 1f;
        Random rand = new Random();
        public bool isPaused = true;
        bool isStarted = true;
        public int isHit = 0;
        public bool isPop = false;

        public Ball(Texture2D newTexture)
        {
            texture = newTexture;
            position = new Vector2((screenWidth - texture.Width) / 2, (screenHeight - texture.Height) / 2);
        }
        
        public void Update(Rectangle playerRect, Rectangle enemyRect)
        {
            if(isStarted && !isPaused)
            {
                RandomLand();
                isStarted = false;
            }

            rectangle = new Rectangle((int)position.X, (int)position.Y, texture.Width, texture.Height);

            if (rectangle.IsOnTopOf(playerRect))
            {
                boost += 0.1f;
                velocity.X = -velocity.X * boost;
                velocity.Y *= boost;
                isPop = true;
            }
            if (rectangle.IsOnTopOf(enemyRect))
            {
                boost += 0.1f;
                velocity.X = -velocity.X * boost;
                velocity.Y *= boost;
                isPop = true;
            }

            position.X += (int)velocity.X;
            position.Y += (int)velocity.Y;

            if (position.X <= -100)
            {
                position.X = (screenWidth - rectangle.Width) / 2;
                position.Y = (screenHeight - rectangle.Height) / 2;
                boost = 1f;
                isPaused = true;
                isStarted = true;
                velocity = Vector2.Zero;
                isHit = 2;
            }
            if (position.Y <= 0)
            {
                velocity.Y = -velocity.Y;
            }
            if (position.X + rectangle.Width >= screenWidth + 100)
            {
                position.X = (screenWidth - rectangle.Width) / 2;
                position.Y = (screenHeight - rectangle.Height) / 2;
                boost = 1f;
                isPaused = true;
                isStarted = true;
                velocity = Vector2.Zero;
                isHit = 1;
            }
            if (position.Y + rectangle.Height >= screenHeight)
            {
                velocity.Y = -velocity.Y;
            }

            Centre = (int)position.Y + (rectangle.Height / 2);
        }

        private void RandomLand()
        {
            int random = rand.Next(0, 4);
            if (random == 0)
            {
                velocity.X = 3f;
                velocity.Y = 3f;
            }
            if (random == 1)
            {
                velocity.X = -3f;
                velocity.Y = 3f;
            }
            if (random == 2)
            {
                velocity.X = -3f;
                velocity.Y = -3f;
            }
            if (random == 3)
            {
                velocity.X = 3f;
                velocity.Y = -3f;
            }
        }
    }
}

static class RectangleHelper
{
    public static bool IsOnTopOf(this Rectangle r1, Rectangle r2)
    {
        return (r1.Bottom >= r2.Top &&
            r1.Top <= r2.Bottom && 
            r1.Right >= r2.Left &&
            r1.Left <= r2.Right);
    }
}